packages/transport/src/chrome-runtime/adapter.ts

```diff
@@ -0,0 +1,86 @@
+  createRouterTransport,
+} from '@connectrpc/connect';
+import { typeRegistry } from '@penumbra-zone/types/src/registry';
+import { iterableToStream, MessageToJson } from '../stream';
+
+interface ChromeRuntimeCallOptions {
+  timeoutMs?: number;
+  signal?: AbortSignal;
+  contextValues?: ContextValues;
+}
+
+/**
+ * This method creates a simple entry function into any service. The entry
+ * accepts a serialized json request, annotated with any type handled by the
+ * service, and returns a serialized single response or stream of responses.
+ *
+ * @param serviceType ServiceType of the service
+ * @param impl Object containing an implementation of the service
+ * @param remote Client to a remote service, for fallback/proxy
+ * @returns (req: JsonValue) => Promise<JsonValue> | ReadableStream<JsonValue>
+ */
+export const adaptServiceImpl = <S extends ServiceType>(
+  serviceType: S,
+  impl: Partial<ServiceImpl<S>>,
+  //ctx?: ContextValues,
+  remote?: PromiseClient<S>,
+) => {
+  const implRoutes = (router: ConnectRouter) => router.service(serviceType, impl);
+  const implClient = createPromiseClient(serviceType, createRouterTransport(implRoutes));
+
+  const namedInfo = new Map(
+    Object.entries(serviceType.methods).map(([localName, methodInfo]) => {
+      return [methodInfo.I.typeName, { localName, ...methodInfo }] as const;
+    }),
+  );
+
+  return (
+    req: JsonValue,
+    opt?: ChromeRuntimeCallOptions,
+  ): Promise<JsonValue | ReadableStream<JsonValue>> => {
+    const unpackedReq = Any.fromJson(req, { typeRegistry }).unpack(typeRegistry)!;
+    const reqType = unpackedReq.getType();
+
+    const info = namedInfo.get(reqType.typeName);
+    if (!info)
+      throw new ConnectError(
+        `Unknown method ${reqType.typeName} on service ${serviceType.typeName}`,
+        ConnectErrorCode.NotFound,
+      );
+    const { localName, kind } = info;
+    const handlerMethod = impl[localName] ? implClient[localName] : remote?.[localName];
+    if (!handlerMethod)
+      throw new ConnectError(
+        `No handler for ${reqType.typeName} of ${serviceType.typeName}`,
+        ConnectErrorCode.NotFound,
+      );
+
+    type AnyUnaryFn = (r: AnyMessage, o?: CallOptions) => Promise<AnyMessage>;
+    type AnyServerStreamingFn = (r: AnyMessage, o?: CallOptions) => AsyncIterable<AnyMessage>;
+    //type AnyClientStreamingFn = (r: AsyncIterable<AnyMessage>, o?: CallOptions) => Promise<AnyMessage>;
+    //type AnyBiDiStreamingFn = (r: AsyncIterable<AnyMessage>, o?: CallOptions) => AsyncIterable<AnyMessage>;
+
+    switch (kind) {
+      case MethodKind.Unary: {
+        const res = (handlerMethod as AnyUnaryFn)(unpackedReq, opt);
+        return res.then(r => Any.pack(r).toJson({ typeRegistry }));
+      }
+      case MethodKind.ServerStreaming: {
+        const res = (handlerMethod as AnyServerStreamingFn)(unpackedReq, opt);
+        return Promise.resolve(iterableToStream(res).pipeThrough(new MessageToJson(typeRegistry)));
```

Comment:
You can convert this return function to an async function and not need the Promise.resolve.

Reply:
the pipeTo creates a promise actually iirc so i will just drop the Promise.resolve